File: packages/javascript/test/location.test.ts
import assert from 'assert';
import { describe, it } from 'node:test';
import {
	LOCATION_LANGUAGES,
	RAND_COUNT_MAX,
	WORD_LANGUAGES,
	randCity,
	randCountry,
	randDistrict,
	randLocation,
	randRegion
} from '../dist/index.js';
import type {
	LocationDetail,
	LocationLanguage,
	LocationLevel,
	WordLanguage
} from '../dist/index.js';
// The datasets are internal, but a location is only real if it is one of theirs —
// these are what tie the output back to the outline each language ships.
import { outline } from '../dist/_internal/parse.js';
import { COUNTRIES } from '../dist/location/data/countries.js';
import { LOCATION_DATA } from '../dist/location/data/index.js';

const SAMPLE = 60;

// What one division's name may be written with. Korean names carry a digit where
// the division does (`종로1가`); US names carry the punctuation their own
// official names do (`O'Fallon`, `Winston-Salem`, `St. Louis`).
const NAME: Record<LocationLanguage, RegExp> = {
	ko: /^[가-힣0-9]+(?: [가-힣0-9]+)?$/,
	en: /^[\p{Script=Latin}0-9 .,'()/-]+$/u
};

// The script a country's name opens on, per word language.
const OPENS: Record<WordLanguage, RegExp> = {
	en: /^\p{Script=Latin}/u,
	ko: /^[가-힣]/,
	ja: /^[぀-ヿ一-鿿]/,
	zh: /^[一-鿿]/,
	vi: /^\p{Script=Latin}/u,
	es: /^\p{Script=Latin}/u,
	it: /^\p{Script=Latin}/u,
	de: /^\p{Script=Latin}/u,
	ru: /^\p{Script=Cyrillic}/u
};

/** The country table as rows of `[code, name, name, …]`. */
const COUNTRY_ROWS = COUNTRIES.table
	.trim()
	.split('\n')
	.map((line: string) => line.trim().split('|'));

/** Every country's name in one language, by its code. */
function countriesIn(language: WordLanguage): Map<string, string> {
	const column = COUNTRIES.languages.indexOf(language) + 1;

	return new Map(COUNTRY_ROWS.map((row: string[]) => [row[0], row[column]]));
}

const DEPTH: Record<LocationLevel, number> = { country: 0, region: 1, city: 2, district: 3 };

type Place = { region: string; city: string | null; district: string | null };

/** Every division a language's outline holds, in the order it holds them. */
function placesOf(language: LocationLanguage): Place[] {
	const data = LOCATION_DATA[language];

	return outline(data.outline, data.levels.length).map((entry) => {
		const place: Place = { region: '', city: null, district: null };

		data.levels.forEach((level, depth) => {
			const name = entry.path[depth] ?? null;

			if (level === 'region') place.region = name ?? '';
			else place[level] = name;
		});

		return place;
	});
}

function keyOf(place: Place): string {
	return [place.region, place.city ?? '', place.district ?? ''].join('/');
}

const LISTED: Record<LocationLanguage, Place[]> = { ko: placesOf('ko'), en: placesOf('en') };

const PLACES: Record<LocationLanguage, Map<string, Place>> = {
	ko: new Map(LISTED.ko.map((place) => [keyOf(place), place])),
	en: new Map(LISTED.en.map((place) => [keyOf(place), place]))
};

/** What a detail writes out, rebuilt from its own levels. */
function written(detail: LocationDetail): string {
	const data = LOCATION_DATA[detail.language];
	const parts = [detail.country, detail.region, detail.city, detail.district].filter(
		(part): part is string => part !== null
	);

	return (data.order === 'largest-first' ? parts : parts.reverse()).join(data.joiner);
}

describe('Location', () => {
	it('randLocation returns one location by default', () => {
		const locations = randLocation();

		assert.strictEqual(locations.length, 1);
		assert.strictEqual(typeof locations[0], 'string');
	});

	it('every generator returns exactly `count` results', () => {
		for (const count of [0, 1, 7, 25]) {
			assert.strictEqual(randLocation({ count }).length, count);
			assert.strictEqual(randCountry({ count }).length, count);
			assert.strictEqual(randRegion({ count }).length, count);
			assert.strictEqual(randCity({ count }).length, count);
			assert.strictEqual(randDistrict({ count }).length, count);
		}

		assert.strictEqual(randLocation({ count: -3 }).length, 0);
		assert.strictEqual(randRegion({ count: RAND_COUNT_MAX + 5 }).length, RAND_COUNT_MAX);
	});

	it('a location is written out from its own levels, in the order its language writes one', () => {
		for (const language of LOCATION_LANGUAGES) {
			for (const detail of randLocation({ language, count: SAMPLE, output: 'detail' })) {
				assert.strictEqual(detail.language, language);
				assert.strictEqual(detail.country, LOCATION_DATA[language].country);
				assert.strictEqual(detail.location, written(detail), detail.location);
			}
		}

		assert.match(randLocation({ language: 'ko' })[0], /^대한민국 /);
		assert.match(randLocation({ language: 'en' })[0], /, United States$/);
	});

	it('every division is one the dataset holds, inside the division written beside it', () => {
		for (const language of LOCATION_LANGUAGES) {
			for (const detail of randLocation({ language, count: SAMPLE * 5, output: 'detail' })) {
				const place = PLACES[language].get(keyOf(detail as Place));

				assert.ok(place, `${language}: ${detail.location} is not in the outline`);

				for (const name of [detail.region, detail.city, detail.district]) {
					if (name !== null) assert.match(name, NAME[language], detail.location);
				}
			}
		}
	});

	it('level says how far down a location goes, and stops at the deepest one its country has', () => {
		for (const level of ['country', 'region', 'city', 'district'] as const) {
			for (const detail of randLocation({
				language: 'ko',
				level,
				count: SAMPLE,
				output: 'detail'
			})) {
				// 세종특별자치시 has no 시·군·구, so a location asked for a city can stop at
				// the region; nothing ever goes past the level asked for.
				assert.ok(DEPTH[detail.level] <= DEPTH[level], detail.location);
				assert.strictEqual(detail.district === null, level !== 'district');
				assert.strictEqual(detail.region === null, level === 'country');
			}
		}

		for (const detail of randLocation({ language: 'en', count: SAMPLE, output: 'detail' })) {
			assert.strictEqual(detail.level, 'city');
			assert.strictEqual(detail.district, null);
		}

		assert.deepStrictEqual(randLocation({ language: 'en', level: 'country', count: 2 }), [
			'United States',
			'United States'
		]);
	});

	it('includeCountry: false writes a location without its country, and the detail keeps it', () => {
		for (const language of LOCATION_LANGUAGES) {
			const data = LOCATION_DATA[language];

			for (const detail of randLocation({
				language,
				includeCountry: false,
				count: SAMPLE,
				output: 'detail'
			})) {
				const parts = [detail.region, detail.city, detail.district].filter(
					(part): part is string => part !== null
				);

				assert.strictEqual(
					detail.location,
					(data.order === 'largest-first' ? parts : parts.reverse()).join(data.joiner)
				);
				assert.strictEqual(detail.country, data.country);
				assert.ok(!detail.location.includes(data.country), detail.location);
			}
		}

		// A location at the country level is the country, so it is written either way.
		assert.deepStrictEqual(
			randLocation({ language: 'en', level: 'country', includeCountry: false }),
			['United States']
		);

		// `startsWith` and the length options read the string that is written.
		for (const location of randLocation({
			language: 'ko',
			includeCountry: false,
			startsWith: '서',
			count: SAMPLE
		})) {
			assert.match(location, /^서울특별시 /);
		}

		for (const location of randLocation({
			language: 'ko',
			includeCountry: false,
			maxLength: 10,
			count: SAMPLE
		})) {
			assert.ok(location.length <= 10, location);
		}
	});

	it('a region with no city is still a location at the city level', () => {
		// Every city-level location of that exact length, which is few enough for the
		// draws `unique` allows to reach all of them.
		const locations = randLocation({
			language: 'ko',
			level: 'city',
			minLength: '대한민국 세종특별자치시'.length,
			maxLength: '대한민국 세종특별자치시'.length,
			unique: true,
			count: 1000
		});

		assert.ok(locations.includes('대한민국 세종특별자치시'));
	});

	it('each level has a generator of its own, and it hands back that level alone', () => {
		const generators = { region: randRegion, city: randCity, district: randDistrict } as const;

		for (const [level, generate] of Object.entries(generators)) {
			for (const detail of generate({ count: SAMPLE, output: 'detail' })) {
				assert.strictEqual(detail.level, level);
				assert.strictEqual(detail.location, detail[level as keyof typeof generators]);
				assert.ok(PLACES[detail.language].has(keyOf(detail as Place)), detail.location);
			}
		}
	});

	it('randCountry names every ISO 3166-1 country, in every word language', () => {
		assert.strictEqual(COUNTRY_ROWS.length, 249);
		assert.strictEqual(new Set(COUNTRY_ROWS.map((row: string[]) => row[0])).size, 249);
		assert.deepStrictEqual([...COUNTRIES.languages].sort(), [...WORD_LANGUAGES].sort());

		for (const row of COUNTRY_ROWS) {
			assert.match(row[0], /^[A-Z]{2}$/);
			assert.strictEqual(row.length, COUNTRIES.languages.length + 1, row[0]);
		}

		for (const language of WORD_LANGUAGES) {
			const named = countriesIn(language);

			for (const detail of randCountry({ language, count: SAMPLE, output: 'detail' })) {
				assert.strictEqual(detail.language, language);
				assert.strictEqual(named.get(detail.code), detail.country, `${language}: ${detail.code}`);
				assert.match(detail.country, OPENS[language], detail.country);
			}

			// Unique by name, and two countries a language names alike would be one.
			assert.strictEqual(
				randCountry({ language, unique: true, count: 300 }).length,
				new Set(named.values()).size
			);
		}
	});

	it('randCountry mixes every word language, not only the ones with divisions', () => {
		const languages = new Set(randCountry({ count: 300, output: 'detail' }).map((d) => d.language));

		assert.deepStrictEqual([...languages].sort(), [...WORD_LANGUAGES].sort());
		assert.deepStrictEqual(
			randCountry({ language: 'de', startsWith: 'Ö', count: 3, unique: true }),
			['Österreich']
		);

		for (const name of randCountry({ language: 'ru', maxLength: 5, count: SAMPLE })) {
			assert.ok(name.length <= 5, name);
		}
	});

	it('a location opens on the name the country table gives its country', () => {
		// `randLocation` and `randCountry` read one table, so they cannot spell a country two ways.
		assert.strictEqual(LOCATION_DATA.ko.country, countriesIn('ko').get('KR'));
		assert.strictEqual(LOCATION_DATA.en.country, countriesIn('en').get('US'));
		assert.deepStrictEqual(randLocation({ language: 'ko', level: 'country' }), ['대한민국']);
	});

	it('a level the country does not have is answered with nothing, and `all` skips that country', () => {
		assert.deepStrictEqual(randDistrict({ language: 'en', count: 5 }), []);

		for (const detail of randDistrict({ count: SAMPLE, output: 'detail' })) {
			assert.strictEqual(detail.language, 'ko');
		}
	});

	it('the mixed language uses every language it knows', () => {
		const languages = new Set(randCity({ count: SAMPLE, output: 'detail' }).map((d) => d.language));

		assert.deepStrictEqual([...languages].sort(), [...LOCATION_LANGUAGES].sort());
	});

	it('an unknown language or level falls back to the default', () => {
		const detail = randLocation({
			language: 'ja' as never,
			level: 'street' as never,
			output: 'detail'
		})[0];

		assert.ok(LOCATION_LANGUAGES.includes(detail.language));
		assert.strictEqual(detail.level === 'district' || detail.language === 'en', true);
	});

	it('startsWith leads every result with the requested character', () => {
		for (const city of randCity({ language: 'en', startsWith: 'Z', count: SAMPLE })) {
			assert.match(city, /^Z/);
		}

		for (const district of randDistrict({ startsWith: '역', count: SAMPLE })) {
			assert.match(district, /^역/);
		}

		// Matched without regard to case, the way every generator matches it.
		assert.match(randRegion({ language: 'en', startsWith: 'n' })[0], /^N/);
	});

	it('a startsWith no language can write, or no division begins with, is answered with nothing', () => {
		assert.deepStrictEqual(randLocation({ language: 'ko', startsWith: 'Q' }), []);
		assert.deepStrictEqual(randCity({ startsWith: 'ж' }), []);
		// Every Korean location opens on the country, so no other character can lead one.
		assert.deepStrictEqual(randLocation({ language: 'ko', startsWith: '서' }), []);
		// And a language that cannot answer is out before a draw, rather than half the draws.
		assert.strictEqual(randCity({ startsWith: 'Z', count: 10 }).length, 10);
	});

	it('results stay inside the requested length range', () => {
		for (const city of randCity({ language: 'en', minLength: 5, maxLength: 6, count: SAMPLE })) {
			assert.ok(city.length >= 5 && city.length <= 6, city);
		}

		for (const location of randLocation({ language: 'ko', maxLength: 16, count: SAMPLE })) {
			assert.ok(location.length <= 16, location);
		}
	});

	it('a range nothing fits is answered with the closest division, not with none', () => {
		const longest = LISTED.en.reduce((most, place) => Math.max(most, place.city?.length ?? 0), 0);

		for (const city of randCity({ language: 'en', minLength: 60, count: 5 })) {
			assert.strictEqual(city.length, longest);
		}

		// An overshoot is worse than an undershoot of the same size.
		for (const region of randRegion({ language: 'ko', maxLength: 2, count: 5 })) {
			assert.strictEqual(region.length, 3, region);
		}
	});

	it('unique never repeats a result, and stops when the pool runs out', () => {
		const regions = randRegion({ language: 'en', unique: true, count: SAMPLE });

		assert.strictEqual(new Set(regions).size, regions.length);
		assert.strictEqual(regions.length, 51);
	});

	it('the Korean dataset stops above the 리, and writes a city district after its city', () => {
		const places = LISTED.ko;
		const regions = places.filter((place) => place.city === null && place.district === null);
		const cities = new Set(places.map((place) => place.city).filter(Boolean));

		assert.strictEqual(regions.length, 16);

		for (const place of places) {
			if (place.district) assert.match(place.district, /(?:동|읍|면|가|로)$/, place.district);
			if (place.district) assert.doesNotMatch(place.district, /리$|출장소/, place.district);
		}

		for (const city of cities) {
			// `수원시장안구` is how the file writes it and not how an address does.
			assert.doesNotMatch(city!, /시\S+구$/, city!);
		}

		assert.ok(cities.has('수원시 장안구'));
		// 세종특별자치시 has no 시·군·구, and its 읍·면·동 sit directly under it.
		assert.ok(
			places.some(
				(place) => place.region === '세종특별자치시' && place.city === null && place.district
			)
		);
		assert.ok(!places.some((place) => place.region === '세종특별자치시' && place.city !== null));
	});

	it('the US dataset is the fifty states and DC, with each place named the way it is called', () => {
		const places = LISTED.en;
		const regions = new Set(places.map((place) => place.region));

		assert.strictEqual(regions.size, 51);
		assert.ok(regions.has('District of Columbia'));
		assert.ok(!regions.has('Puerto Rico'));

		const seen = new Set<string>();

		for (const place of places) {
			if (!place.city) continue;

			assert.doesNotMatch(
				place.city,
				/ (?:city|town|village|borough|CDP|municipality)$|\(balance\)/,
				place.city
			);

			const key = `${place.region}/${place.city}`;

			assert.ok(!seen.has(key), `${key} is listed twice`);
			seen.add(key);
		}
	});
});

File: packages/dart/test/location_test.dart
import 'package:randino/randino.dart';
// The datasets are internal, but a location is only real if it is one of theirs —
// these are what tie the output back to the outline each language ships.
import 'package:randino/src/internal/parse.dart';
import 'package:randino/src/location/data/index.dart';
import 'package:randino/src/location/data/types.dart';
import 'package:test/test.dart';

const int sample = 60;

// What one division's name may be written with. Korean names carry a digit where
// the division does (`종로1가`); US names carry the punctuation their own
// official names do (`O'Fallon`, `Winston-Salem`, `St. Louis`).
final Map<LocationLanguage, RegExp> nameScript = <LocationLanguage, RegExp>{
  LocationLanguage.ko: RegExp(r'^[가-힣0-9]+(?: [가-힣0-9]+)?$'),
  LocationLanguage.en: RegExp(r"^[\p{Script=Latin}0-9 .,'()/-]+$", unicode: true),
};

const Map<LocationLevel, int> depthOf = <LocationLevel, int>{
  LocationLevel.country: 0,
  LocationLevel.region: 1,
  LocationLevel.city: 2,
  LocationLevel.district: 3,
};

/// One division the outline holds, by the names of every level down to it.
typedef Place = ({String region, String? city, String? district});

/// Every division a language's outline holds, in the order it holds them.
List<Place> placesOf(LocationLanguage language) {
  final data = locationData[language]!;

  return <Place>[
    for (final entry in outline(data.outline, data.levels.length))
      () {
        final named = <LocationLevel, String?>{};

        for (var depth = 0; depth < data.levels.length; depth += 1) {
          named[data.levels[depth]] = depth < entry.path.length ? entry.path[depth] : null;
        }

        return (
          region: named[LocationLevel.region] ?? '',
          city: named[LocationLevel.city],
          district: named[LocationLevel.district],
        );
      }(),
  ];
}

String keyOf(String? region, String? city, String? district) =>
    <String>[region ?? '', city ?? '', district ?? ''].join('/');

String placeKey(Place place) => keyOf(place.region, place.city, place.district);

String detailKey(LocationDetail detail) => keyOf(detail.region, detail.city, detail.district);

final Map<LocationLanguage, List<Place>> listed = <LocationLanguage, List<Place>>{
  for (final language in locationLanguages) language: placesOf(language),
};

final Map<LocationLanguage, Map<String, Place>> places = <LocationLanguage, Map<String, Place>>{
  for (final language in locationLanguages)
    language: <String, Place>{for (final place in listed[language]!) placeKey(place): place},
};

/// What a detail writes out, rebuilt from its own levels.
String written(LocationDetail detail) {
  final data = locationData[detail.language]!;
  final parts =
      <String?>[detail.country, detail.region, detail.city, detail.district].nonNulls.toList();

  return (data.order == LocationOrder.largestFirst ? parts : parts.reversed).join(data.joiner);
}

void main() {
  group('Location', () {
    test('randLocation returns one location by default', () {
      final locations = randLocation();

      expect(locations, hasLength(1));
      expect(locations.first, isNotEmpty);
    });

    test('every generator returns exactly `count` results', () {
      for (final count in <int>[0, 1, 7, 25]) {
        expect(randLocation(count: count), hasLength(count));
        expect(randCountry(count: count), hasLength(count));
        expect(randRegion(count: count), hasLength(count));
        expect(randCity(count: count), hasLength(count));
        expect(randDistrict(count: count), hasLength(count));
      }

      expect(randLocation(count: -3), isEmpty);
      expect(randRegion(count: randCountMax + 5), hasLength(randCountMax));
    });

    test('a location is written out from its own levels, in the order its language writes one', () {
      for (final language in locationLanguages) {
        for (final detail in randLocationDetails(language: language, count: sample)) {
          expect(detail.language, language);
          expect(detail.country, locationData[language]!.country);
          expect(detail.location, written(detail), reason: detail.location);
        }
      }

      expect(randLocation(language: LocationLanguage.ko).first, startsWith('대한민국 '));
      expect(randLocation(language: LocationLanguage.en).first, endsWith(', United States'));
    });

    test('every division is one the dataset holds, inside the division written beside it', () {
      for (final language in locationLanguages) {
        for (final detail in randLocationDetails(language: language, count: sample * 5)) {
          expect(
            places[language]!.containsKey(detailKey(detail)),
            isTrue,
            reason: '$language: ${detail.location} is not in the outline',
          );

          for (final name in <String?>[detail.region, detail.city, detail.district].nonNulls) {
            expect(name, matches(nameScript[language]!), reason: detail.location);
          }
        }
      }
    });

    test(
      'level says how far down a location goes, and stops at the deepest one its country has',
      () {
        for (final level in LocationLevel.values) {
          for (final detail in randLocationDetails(
            language: LocationLanguage.ko,
            level: level,
            count: sample,
          )) {
            // 세종특별자치시 has no 시·군·구, so a location asked for a city can stop at
            // the region; nothing ever goes past the level asked for.
            expect(
              depthOf[detail.level]!,
              lessThanOrEqualTo(depthOf[level]!),
              reason: detail.location,
            );
            expect(
              detail.district == null,
              level != LocationLevel.district,
              reason: detail.location,
            );
            expect(detail.region == null, level == LocationLevel.country, reason: detail.location);
          }
        }

        for (final detail in randLocationDetails(language: LocationLanguage.en, count: sample)) {
          expect(detail.level, LocationLevel.city);
          expect(detail.district, isNull);
        }

        expect(
          randLocation(language: LocationLanguage.en, level: LocationLevel.country, count: 2),
          <String>['United States', 'United States'],
        );
      },
    );

    test('a region with no city is still a location at the city level', () {
      // Every city-level location of that exact length, which is few enough for the
      // draws `unique` allows to reach all of them.
      final locations = randLocation(
        language: LocationLanguage.ko,
        level: LocationLevel.city,
        minLength: '대한민국 세종특별자치시'.length,
        maxLength: '대한민국 세종특별자치시'.length,
        unique: true,
        count: 1000,
      );

      expect(locations, contains('대한민국 세종특별자치시'));
    });

    test('each level has a generator of its own, and it hands back that level alone', () {
      final generators =
          <(LocationLevel, List<LocationDetail> Function(), String? Function(LocationDetail))>[
            (LocationLevel.region, () => randRegionDetails(count: sample), (d) => d.region),
            (LocationLevel.city, () => randCityDetails(count: sample), (d) => d.city),
            (LocationLevel.district, () => randDistrictDetails(count: sample), (d) => d.district),
          ];

      for (final (level, generate, named) in generators) {
        for (final detail in generate()) {
          expect(detail.level, level);
          expect(detail.location, named(detail));
          expect(
            places[detail.language]!.containsKey(detailKey(detail)),
            isTrue,
            reason: detail.location,
          );
        }
      }

      expect(randCountry(language: LocationLanguage.ko), <String>['대한민국']);

      final countries = randCountryDetails(language: LocationLanguage.en);

      expect(countries, hasLength(1));
      expect(countries.first.location, 'United States');
      expect(countries.first.language, LocationLanguage.en);
      expect(countries.first.level, LocationLevel.country);
      expect(countries.first.country, 'United States');
      expect(countries.first.region, isNull);
      expect(countries.first.city, isNull);
      expect(countries.first.district, isNull);
    });

    test(
      'a level the country does not have is answered with nothing, and `all` skips that country',
      () {
        expect(randDistrict(language: LocationLanguage.en, count: 5), isEmpty);

        for (final detail in randDistrictDetails(count: sample)) {
          expect(detail.language, LocationLanguage.ko);
        }
      },
    );

    test('the mixed language uses every language it knows', () {
      final languages = randCityDetails(count: sample).map((detail) => detail.language).toSet();

      expect(languages, locationLanguages.toSet());
    });

    test('a null language and the default level are every language and the district', () {
      // The npm package's `language: 'ja'` and `level: 'street'` fall back to these
      // two; Dart's enums cannot be handed an unknown value, so what is left to check
      // is that leaving both out means the same.
      final detail = randLocationDetails().first;

      expect(locationLanguages, contains(detail.language));
      expect(
        detail.level == LocationLevel.district || detail.language == LocationLanguage.en,
        isTrue,
      );
    });

    test('startsWith leads every result with the requested character', () {
      for (final city in randCity(language: LocationLanguage.en, startsWith: 'Z', count: sample)) {
        expect(city, startsWith('Z'));
      }

      for (final district in randDistrict(startsWith: '역', count: sample)) {
        expect(district, startsWith('역'));
      }

      // Matched without regard to case, the way every generator matches it.
      expect(randRegion(language: LocationLanguage.en, startsWith: 'n').first, startsWith('N'));
    });

    test(
      'a startsWith no language can write, or no division begins with, is answered with nothing',
      () {
        expect(randLocation(language: LocationLanguage.ko, startsWith: 'Q'), isEmpty);
        expect(randCity(startsWith: 'ж'), isEmpty);
        // Every Korean location opens on the country, so no other character can lead one.
        expect(randLocation(language: LocationLanguage.ko, startsWith: '서'), isEmpty);
        // And a language that cannot answer is out before a draw, rather than half the draws.
        expect(randCity(startsWith: 'Z', count: 10), hasLength(10));
      },
    );

    test('results stay inside the requested length range', () {
      for (final city in randCity(
        language: LocationLanguage.en,
        minLength: 5,
        maxLength: 6,
        count: sample,
      )) {
        expect(city.length, inInclusiveRange(5, 6), reason: city);
      }

      for (final location in randLocation(
        language: LocationLanguage.ko,
        maxLength: 16,
        count: sample,
      )) {
        expect(location.length, lessThanOrEqualTo(16), reason: location);
      }
    });

    test('a range nothing fits is answered with the closest division, not with none', () {
      final longest = listed[LocationLanguage.en]!.fold<int>(
        0,
        (most, place) =>
            place.city != null && place.city!.length > most ? place.city!.length : most,
      );

      for (final city in randCity(language: LocationLanguage.en, minLength: 60, count: 5)) {
        expect(city.length, longest);
      }

      // An overshoot is worse than an undershoot of the same size.
      for (final region in randRegion(language: LocationLanguage.ko, maxLength: 2, count: 5)) {
        expect(region.length, 3, reason: region);
      }
    });

    test('unique never repeats a result, and stops when the pool runs out', () {
      final regions = randRegion(language: LocationLanguage.en, unique: true, count: sample);

      expect(regions.toSet(), hasLength(regions.length));
      expect(regions, hasLength(51));
      expect(randCountry(unique: true, count: 5), hasLength(2));
    });

    test('the Korean dataset stops above the 리, and writes a city district after its city', () {
      final korean = listed[LocationLanguage.ko]!;
      final regions = korean.where((place) => place.city == null && place.district == null);
      final cities = korean.map((place) => place.city).nonNulls.toSet();

      expect(regions, hasLength(16));

      for (final place in korean) {
        final district = place.district;

        if (district != null) {
          expect(district, matches(RegExp(r'(?:동|읍|면|가|로)$')), reason: district);
          expect(district, isNot(matches(RegExp(r'리$|출장소'))), reason: district);
        }
      }

      for (final city in cities) {
        // `수원시장안구` is how the file writes it and not how an address does.
        expect(city, isNot(matches(RegExp(r'시\S+구$'))), reason: city);
      }

      expect(cities, contains('수원시 장안구'));
      // 세종특별자치시 has no 시·군·구, and its 읍·면·동 sit directly under it.
      expect(
        korean.any(
          (place) => place.region == '세종특별자치시' && place.city == null && place.district != null,
        ),
        isTrue,
      );
      expect(korean.any((place) => place.region == '세종특별자치시' && place.city != null), isFalse);
    });

    test(
      'the US dataset is the fifty states and DC, with each place named the way it is called',
      () {
        final american = listed[LocationLanguage.en]!;
        final regions = american.map((place) => place.region).toSet();

        expect(regions, hasLength(51));
        expect(regions, contains('District of Columbia'));
        expect(regions, isNot(contains('Puerto Rico')));

        final seen = <String>{};
        final suffixed = RegExp(r' (?:city|town|village|borough|CDP|municipality)$|\(balance\)');

        for (final place in american) {
          final city = place.city;

          if (city == null) continue;

          expect(city, isNot(matches(suffixed)), reason: city);

          final key = '${place.region}/$city';

          expect(seen.contains(key), isFalse, reason: '$key is listed twice');
          seen.add(key);
        }
      },
    );
  });
}
